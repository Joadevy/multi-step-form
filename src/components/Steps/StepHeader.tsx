import React, { FC } from "react";

type props = {
  steps: number[];
  active: number;
};

const stepDescription = ["YOUR INFO", "SELECT PLAN", "ADDS-ON", "SUMMARY"];

export const StepHeader: FC<props> = ({ steps, active }) => {
  return (
    <header className="h-48 w-full absolute lg:w-[300px] lg:relative lg:h-auto top-0 -z-10 bg-mobileHero lg:bg-desktopHero flex bg-cover lg:bg-contain bg-no-repeat">
      <ul className="flex lg:flex-col gap-4 lg:gap-8 w-full justify-center lg:justify-start pt-10 lg:pl-10 text-white">
        {steps.map((step) => (
          <li key={step} className="lg:flex lg:gap-2 lg:items-center">
            <div
              className={
                "rounded-full border border-white h-9 w-9 lg:h-8 lg:w-8 flex justify-center items-center " +
                (active === step ? "bg-white text-p-marine-blue" : "")
              }
            >
              <p>{step}</p>
            </div>
            <div className="hidden lg:block text-white text-sm lg:text-xs">
              <p className=" text-p-pastel-blue">{`STEP ${step}`}</p>
              <p className="font-bold">{stepDescription[step - 1]}</p>
            </div>
          </li>
        ))}
      </ul>
    </header>
  );
};
