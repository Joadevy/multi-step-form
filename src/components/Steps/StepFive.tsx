import { FC } from "react";

import thanksImg from "../../assets/images/icon-thank-you.svg";

import { StepHeader } from "./StepHeader";

export const StepFive: FC = () => {
  return (
    <div className="lg:flex lg:my-10 lg:mx-48 gap-5">
      <StepHeader active={4} steps={[1, 2, 3, 4]} />

      <div
        className="flex flex-col rounded-xl mx-5 my-28 p-8 bg-white 
       lg:mx-0 lg:my-0 lg:p-0 lg:w-[400px] lg:h-[500px] lg:min-h-0 relative lg:items-center lg:justify-center "
      >
        <header className="flex flex-col mt-10 gap-3 items-center">
          <div className="w-20 h-20 mb-3">
            <img alt="Thanks for your order!" src={thanksImg} />
          </div>
          <h2 className="font-bold text-3xl text-p-marine-blue">Thank you!</h2>
          <p className="text-n-cool-gray text-xl font-normal text-center lg:text-base">
            Thanks for confirming your subscription! We hope you have fun using
            our platform. If you ever need support, please feel free to email us
            at support@loremgaming.com.
          </p>
        </header>
      </div>
    </div>
  );
};
