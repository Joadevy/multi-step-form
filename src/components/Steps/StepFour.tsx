import { FC } from "react";

import { User } from "../Form";

import { StepHeader } from "./StepHeader";

type props = {
  userData: User;
  prevStep: () => void;
  nextStep: () => void;
  backToPlan: () => void;

  handleUserData: (_: string, __: Map<string, number> | number) => void;
};

const capitalizeFirst = (word: string) => {
  return word.charAt(0).toUpperCase() + word.slice(1);
};

export const StepFour: FC<props> = ({
  userData,
  handleUserData,
  backToPlan,
  prevStep,
  nextStep,
}) => {
  const submitUserData = (e: any) => {
    e.preventDefault();

    // checking if value of first name and last name is empty show error else take to step 2
    // if (!isValid(values.name) || !isValid(values.phone) || !isValid(values.email)) {
    //   setError(true);
    // } else {
    //   nextStep();
    // }

    if (userData.plan) {
      nextStep();
    }
  };

  const getTotalPrice = (): number => {
    let total = 0;

    total += userData.pricePlan;
    userData.addsOn.forEach((price: number) => {
      total += price;
    });

    return total;
  };

  return (
    <div>
      <StepHeader active={4} steps={[1, 2, 3, 4]} />

      <div className="flex flex-col rounded-xl mx-5 my-24 p-8 bg-white">
        <header className="flex flex-col gap-2">
          <h2 className="font-bold text-3xl text-p-marine-blue">
            Finishing up
          </h2>
          <p className="text-n-cool-gray text-lg font-medium">
            Double-check everything looks OK before confirming
          </p>
        </header>
        <form className="flex flex-col gap-4 mt-6" onSubmit={submitUserData}>
          <ul className="flex flex-col gap-4">
            <li className="flex justify-between items-center mb-3">
              <div>
                <h3 className="font-bold text-p-marine-blue text-lg">{`${capitalizeFirst(
                  userData.plan
                )} (${capitalizeFirst(userData.planDuration)})`}</h3>
                <button onClick={backToPlan}>
                  <p className="underline text-n-cool-gray text-lg">Change</p>
                </button>
              </div>
              <p className="font-bold text-p-marine-blue text-lg mr-2">
                {userData.planDuration === "monthly"
                  ? `$${userData.pricePlan}/mo`
                  : `$${userData.pricePlan}/yr`}
              </p>
            </li>

            {Array.from(userData.addsOn).map(
              (addOn: [string, number], index: number) => (
                <li key={index} className="flex justify-between items-center ">
                  <div>
                    <h3 className="font-medium text-n-cool-gray text-lg">
                      {capitalizeFirst(
                        addOn[0] as
                          | "online service"
                          | "larger storage"
                          | "customizable profile"
                      )}
                    </h3>
                  </div>
                  <p className="font-bold text-p-marine-blue text-lg mr-2">
                    {userData.planDuration === "monthly"
                      ? `+$${addOn[1] as number}/mo`
                      : `+$${addOn[1] as number}/yr`}
                  </p>
                </li>
              )
            )}

            <li className="flex justify-between items-center mt-5">
              <div>
                <h3 className="font-medium text-n-cool-gray text-lg">{`${capitalizeFirst(
                  "total"
                )} (${
                  userData.planDuration === "monthly" ? "per month" : "per year"
                })`}</h3>
              </div>
              <p className="font-bold text-p-purplish-blue text-xl mr-2">
                {userData.planDuration === "monthly"
                  ? `$${getTotalPrice()}/mo`
                  : `$${getTotalPrice()}/yr`}
              </p>
            </li>
          </ul>
          <button
            className="absolute z-20 bottom-6 left-6 text-n-cool-gray text-lg hover:text-p-marine-blue transition-colors"
            onClick={prevStep}
          >
            Go Back
          </button>
          <input
            className="absolute z-20 bottom-4 right-6 w-3/12 rounded-md px-4 py-3 bg-p-purplish-blue text-white font-bold cursor-pointer hover:opacity-70 transition-colors"
            type="submit"
            value={"Confirm"}
          />
        </form>
      </div>
    </div>
  );
};
