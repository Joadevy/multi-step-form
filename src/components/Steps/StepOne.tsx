import { ChangeEventHandler, FC, useState } from "react";

import { StepHeader } from "./StepHeader";

type props = {
  values: any;
  nextStep: () => void;
  handleUserData: (_: any) => ChangeEventHandler;
};

export const StepOne: FC<props> = ({ values, nextStep, handleUserData }) => {
  const [error, setError] = useState(false);
  const submitFormData = (e: any) => {
    e.preventDefault();

    // checking if value of first name and last name is empty show error else take to step 2
    // if (!isValid(values.name) || !isValid(values.phone) || !isValid(values.email)) {
    //   setError(true);
    // } else {
    //   nextStep();
    // }

    nextStep();
  };

  return (
    <div>
      <StepHeader active={1} steps={[1, 2, 3, 4]} />
      <div className="flex flex-col rounded-xl mx-5 my-24 p-8 bg-white">
        <header className="flex flex-col gap-2">
          <h2 className="font-bold text-3xl text-p-marine-blue">
            Personal info
          </h2>
          <p className="text-n-cool-gray text-lg font-medium">
            Please provide your name, email adress, and phone number.
          </p>
        </header>
        <form className="flex flex-col gap-4 mt-6" onSubmit={submitFormData}>
          <label
            className="text-p-marine-blue font-medium text-base flex flex-col gap-1"
            htmlFor="name"
          >
            Name
            <input
              required
              className="rounded-sm border-2 border-n-light-gray p-1"
              defaultValue={values.name}
              id="name"
              placeholder="e.g. Stephen King"
              type="text"
              onChange={handleUserData("name")}
            />
          </label>

          <label
            className="text-p-marine-blue font-medium text-base flex flex-col gap-1"
            htmlFor="email"
          >
            Email Address
            <input
              required
              className="rounded-sm border-2 border-n-light-gray p-1"
              defaultValue={values.email}
              id="email"
              placeholder="e.g. stephenking@lorem.com"
              type="text"
              onChange={handleUserData("email")}
            />
          </label>

          <label
            className="text-p-marine-blue font-medium text-base flex flex-col gap-1"
            htmlFor="phone"
          >
            Phone Number
            <input
              required
              className="rounded-sm border-2 border-n-light-gray p-1"
              defaultValue={values.phone}
              id="phone"
              placeholder="e.g. +1 234 567 890"
              type="text"
              onChange={handleUserData("phone")}
            />
          </label>

          <input
            className="absolute z-20 bottom-4 right-6 w-3/12 rounded-md px-4 py-3 bg-p-marine-blue text-white font-bold cursor-pointer hover:opacity-50"
            type="submit"
            value={"Next Step"}
          />
        </form>
      </div>
    </div>
  );
};
