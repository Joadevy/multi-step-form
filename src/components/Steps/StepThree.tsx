import { FC } from "react";

import { AddOn } from "../AddOn";
import { addsOnPricesMonthly, addsOnPricesYearly } from "../../Prices";
import { User } from "../Form";
import { GrayButton } from "../Buttons/GrayButton";
import { NextButton } from "../Buttons/NextButton";
import { Container } from "../Container";
import { Header } from "../Header";

import { StepHeader } from "./StepHeader";

type props = {
  userData: User;
  prevStep: () => void;
  nextStep: () => void;
  handleUserData: (_: string, __: Map<string, number> | number) => void;
};

export const StepThree: FC<props> = ({
  userData,
  prevStep,
  nextStep,
  handleUserData,
}) => {
  const submitUserData = (e: any) => {
    e.preventDefault();

    if (userData.plan) {
      updatePrices();
      nextStep();
    }
  };

  const handleAdd = (title: string, price: number) => {
    const addsOn = new Map(userData.addsOn as Map<string, number>);

    const exist = addsOn.has(title);

    if (!exist) {
      addsOn.set(title, price);
      handleUserData("addsOn", addsOn);
    }
  };

  const updatePrices = () => {
    const draft: Map<string, number> = new Map(userData.addsOn);

    userData.planDuration === "monthly"
      ? draft.forEach((_, addOn) => {
          draft.set(
            addOn,
            addsOnPricesMonthly[
              addOn as
                | "online service"
                | "larger storage"
                | "customizable profile"
            ]
          );
        })
      : draft.forEach((_, addOn) =>
          draft.set(
            addOn,
            addsOnPricesYearly[
              addOn as
                | "online service"
                | "larger storage"
                | "customizable profile"
            ]
          )
        );

    handleUserData("addsOn", draft);
  };

  const handleRemove = (title: string) => {
    const addsOn = new Map(userData.addsOn as Map<string, number>);
    const exist = addsOn.has(title);

    if (exist) {
      addsOn.delete(title);
      handleUserData("addsOn", addsOn);
    }
  };

  return (
    <div className="lg:flex lg:my-10 lg:mx-48 gap-5 lg:p-4 lg:shadow-md bg-white lg:rounded-lg">
      <StepHeader active={3} steps={[1, 2, 3, 4]} />
      <div className="min-h-screen lg:w-[400px] lg:h-[500px] 2xl:w-[700px] 2xl:h-[800px]  lg:min-h-0 relative lg:p-4  ">
        <Container>
          <Header
            desc="Add-ons help enhace your gaming experience."
            title="Pick add-ons"
          />

          <form className="flex flex-col gap-4 mt-6" onSubmit={submitUserData}>
            <ul className="flex flex-col gap-5">
              <AddOn
                checked={userData.addsOn.has("online service")}
                desc={"Access to multiplayer games"}
                handleAdd={handleAdd}
                handleRemove={handleRemove}
                price={userData.planDuration === "monthly" ? 1 : 10}
                title={"Online service"}
                type={userData.planDuration}
              />
              <AddOn
                checked={userData.addsOn.has("larger storage")}
                desc={"Extra 1TB of cloud save"}
                handleAdd={handleAdd}
                handleRemove={handleRemove}
                price={userData.planDuration === "monthly" ? 2 : 20}
                title={"Larger storage"}
                type={userData.planDuration}
              />
              <AddOn
                checked={userData.addsOn.has("customizable profile")}
                desc={"Custom theme on your profile"}
                handleAdd={handleAdd}
                handleRemove={handleRemove}
                price={userData.planDuration === "monthly" ? 2 : 20}
                title={"Customizable profile"}
                type={userData.planDuration}
              />
            </ul>
            <GrayButton action={prevStep} label={"Go Back"} />
            <NextButton bgColor="bg-p-marine-blue" label="Next Step" />
          </form>
        </Container>
      </div>
    </div>
  );
};
