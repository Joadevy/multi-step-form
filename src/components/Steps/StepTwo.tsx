import { FC } from "react";

import { Plan } from "../Plan";
import arcadeImg from "../../assets/images/icon-arcade.svg";
import advancedImg from "../../assets/images/icon-advanced.svg";
import proImg from "../../assets/images/icon-pro.svg";
import Switcher from "../Switcher";
import { planPricesMonthly, planPricesYearly } from "../../Prices";
import { User } from "../Form";

import { StepHeader } from "./StepHeader";

type props = {
  userData: User;
  nextStep: () => void;
  prevStep: () => void;
  handleUserData: (_: string, __: string | number) => void;
};

export const StepTwo: FC<props> = ({
  userData,
  nextStep,
  prevStep,
  handleUserData,
}) => {
  const submitUserData = (e: any) => {
    e.preventDefault();

    // checking if value of first name and last name is empty show error else take to step 2
    // if (!isValid(values.name) || !isValid(values.phone) || !isValid(values.email)) {
    //   setError(true);
    // } else {
    //   nextStep();
    // }

    if (userData.plan) {
      switchPrices();
      nextStep();
    }
  };

  const isActive = (plan: string) => {
    return plan === userData.plan;
  };

  const handlePlanDuration = () => {
    userData.planDuration === "monthly"
      ? handleUserData("planDuration", "yearly")
      : handleUserData("planDuration", "monthly");
  };

  const switchPrices = () => {
    userData.planDuration === "monthly"
      ? handleUserData(
          "pricePlan",
          planPricesMonthly[userData.plan as "arcade" | "advanced" | "pro"]
        )
      : handleUserData(
          "pricePlan",
          planPricesYearly[userData.plan as "arcade" | "advanced" | "pro"]
        );
  };

  return (
    <div>
      <StepHeader active={2} steps={[1, 2, 3, 4]} />
      <div className="flex flex-col rounded-xl mx-5 my-28 p-8 bg-white">
        <header className="flex flex-col gap-2">
          <h2 className="font-bold text-3xl text-p-marine-blue">
            Select your plan
          </h2>
          <p className="text-n-cool-gray text-lg font-medium">
            You have the option of monthly or yearly biling.
          </p>
        </header>
        <form className="flex flex-col gap-4 mt-6" onSubmit={submitUserData}>
          <Plan
            active={isActive("arcade")}
            handlePlan={handleUserData}
            img={arcadeImg}
            monthsFree={2}
            name={"Arcade"}
            price={userData.planDuration === "monthly" ? 9 : 90}
            type={userData.planDuration}
          />
          <Plan
            active={isActive("advanced")}
            handlePlan={handleUserData}
            img={advancedImg}
            monthsFree={2}
            name={"Advanced"}
            price={userData.planDuration === "monthly" ? 12 : 120}
            type={userData.planDuration}
          />
          <Plan
            active={isActive("pro")}
            handlePlan={handleUserData}
            img={proImg}
            monthsFree={2}
            name={"Pro"}
            price={userData.planDuration === "monthly" ? 15 : 150}
            type={userData.planDuration}
          />
          <div className="flex gap-4 absolute bottom-28 self-center">
            <p
              className={
                userData.planDuration === "monthly"
                  ? "text-p-marine-blue font-bold"
                  : " text-n-cool-gray font-bold"
              }
            >
              Monthly
            </p>
            <Switcher handlerStatus={handlePlanDuration} />
            <p
              className={
                userData.planDuration === "yearly"
                  ? "text-p-marine-blue font-bold"
                  : " text-n-cool-gray font-bold"
              }
            >
              Yearly
            </p>
          </div>
          <button
            className="absolute z-20 bottom-6 left-6 text-n-cool-gray text-lg hover:text-p-marine-blue transition-colors"
            onClick={prevStep}
          >
            Go Back
          </button>
          <input
            className="absolute z-20 bottom-4 right-6 w-3/12 rounded-md px-4 py-3 bg-p-marine-blue text-white font-bold cursor-pointer hover:opacity-70 transition-colors"
            type="submit"
            value={"Next Step"}
          />
        </form>
      </div>
    </div>
  );
};
