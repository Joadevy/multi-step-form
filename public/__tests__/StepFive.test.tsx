import { test, describe, expect } from "vitest";
import { render } from "@testing-library/react";

import { StepFive } from "../../src/components/Steps/StepFive";

describe("Step Five", async () => {
  test("should render correctly with the thanks information", async () => {
    const { asFragment } = render(<StepFive />);

    expect(asFragment()).toMatchSnapshot();
  });
});
