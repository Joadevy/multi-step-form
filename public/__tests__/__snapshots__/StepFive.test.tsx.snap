// Vitest Snapshot v1

exports[`Step Five > should render correctly with the thanks information 1`] = `
<DocumentFragment>
  <div
    class="max-h-screen lg:flex lg:my-10 lg:mx-48 gap-5 lg:p-4 lg:shadow-md bg-white lg:rounded-lg"
  >
    <header
      class="h-[15vh] w-full absolute lg:z-50 lg:w-[275px] 2xl:w-[400px] lg:relative lg:h-auto top-0 -z-10 bg-mobileHero lg:bg-desktopHero flex bg-cover lg:bg-contain bg-no-repeat"
    >
      <ul
        class="flex lg:flex-col gap-4 lg:gap-8 2xl-gap-10 w-full justify-center lg:justify-start pt-10 lg:pl-10 text-white"
      >
        <li
          class="lg:flex lg:gap-2 lg:items-center"
        >
          <div
            class="rounded-full border border-white h-9 w-9 lg:h-8 lg:w-8 flex justify-center items-center "
          >
            <p>
              1
            </p>
          </div>
          <div
            class="hidden lg:block text-white text-sm lg:text-xs"
          >
            <p
              class=" text-p-pastel-blue"
            >
              STEP 1
            </p>
            <p
              class="font-bold"
            >
              YOUR INFO
            </p>
          </div>
        </li>
        <li
          class="lg:flex lg:gap-2 lg:items-center"
        >
          <div
            class="rounded-full border border-white h-9 w-9 lg:h-8 lg:w-8 flex justify-center items-center "
          >
            <p>
              2
            </p>
          </div>
          <div
            class="hidden lg:block text-white text-sm lg:text-xs"
          >
            <p
              class=" text-p-pastel-blue"
            >
              STEP 2
            </p>
            <p
              class="font-bold"
            >
              SELECT PLAN
            </p>
          </div>
        </li>
        <li
          class="lg:flex lg:gap-2 lg:items-center"
        >
          <div
            class="rounded-full border border-white h-9 w-9 lg:h-8 lg:w-8 flex justify-center items-center "
          >
            <p>
              3
            </p>
          </div>
          <div
            class="hidden lg:block text-white text-sm lg:text-xs"
          >
            <p
              class=" text-p-pastel-blue"
            >
              STEP 3
            </p>
            <p
              class="font-bold"
            >
              ADDS-ON
            </p>
          </div>
        </li>
        <li
          class="lg:flex lg:gap-2 lg:items-center"
        >
          <div
            class="rounded-full border border-white h-9 w-9 lg:h-8 lg:w-8 flex justify-center items-center bg-white text-p-marine-blue"
          >
            <p>
              4
            </p>
          </div>
          <div
            class="hidden lg:block text-white text-sm lg:text-xs"
          >
            <p
              class=" text-p-pastel-blue"
            >
              STEP 4
            </p>
            <p
              class="font-bold"
            >
              SUMMARY
            </p>
          </div>
        </li>
      </ul>
    </header>
    <div
      class="flex flex-col rounded-xl mx-5 my-[15vh] p-8 bg-white 
       lg:mx-0 lg:my-0 lg:p-0 lg:w-[400px] lg:h-[500px] 2xl:w-[700px] 2xl:h-[800px]  lg:min-h-0 relative lg:items-center lg:justify-center "
    >
      <header
        class="flex flex-col mt-10 gap-3 items-center lg:px-14 lg:mt-0"
      >
        <div
          class="w-20 h-20 mb-3"
        >
          <img
            alt="Thanks for your order!"
            src="/src/assets/images/icon-thank-you.svg"
          />
        </div>
        <h2
          class="font-bold text-3xl text-p-marine-blue"
        >
          Thank you!
        </h2>
        <p
          class="text-n-cool-gray text-xl font-normal text-center lg:text-base"
        >
          Thanks for confirming your subscription! We hope you have fun using our platform. If you ever need support, please feel free to email us at support@loremgaming.com.
        </p>
      </header>
    </div>
  </div>
</DocumentFragment>
`;
